streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -93,10 +93,42 @@
     private static final String JMX_PREFIX = "kafka.streams";
 
     // container states
-    private static final int CREATED = 0;
-    private static final int RUNNING = 1;
-    private static final int STOPPED = 2;
-    private int state = CREATED;
+    /**
+     * Kafka Streams states are the possible state that a Kafka Streams instance can be in.
+     * An instance should only be in one state at a time.
+     * Note this instance will be in "Rebalancing" state if any of its threads is rebalancing
+     * The expected state transition with the following defined states is:
+     *
+     *                +-----------+
+     *         +<-----|Created    |
+     *         |      +-----+-----+
+     *         |             |
+     *         |             v
+     *         |       +-----+---------+
+     *         +<----- | Rebalancing   |<--------+
+     *         |       +-----+---------+         ^
+     *         |             |                   |
+     *         |             v                   |
+     *         |       +-----+------------+      |
+     *         |       |Running           |------+
+     *         |       +-----+------------+
+     *         |             |
+     *         |             v
+     *         |     +-----+----------+
+     *         +---->|Pending         |
+     *               |Shutdown        |
+     *               +-----+----------+
+     *                     |
+     *                     v
+     *               +-----+-----+
+     *               |Not Running|
+     *               +-----------+
+     *
+     * Custom states is also allowed for cases where there are custom kafka states for different scenarios.
+     */
+    public enum KafkaStreamsState { CREATED, RUNNING, REBALANCING, PENDING_SHUTDOWN, NOT_RUNNING }
```

Comment:
Since this is an inner enum, you could simply call it `State` and users would reference it as `KafkaStreams.State`